src/libsystemd/sd-device/sd-device.c

```diff
@@ -245,6 +247,36 @@ _public_ int sd_device_new_from_devnum(sd_device **ret, char type, dev_t devnum)
         return sd_device_new_from_syspath(ret, syspath);
 }
 
+_public_ int sd_device_new_from_ifname(sd_device **ret, const char *ifname) {
+        int r;
+
+        assert_return(ret, -EINVAL);
+        assert_return(ifname, -EINVAL);
+
+        /* This also accepts alternative interface names. Please use
+         * sd_device_new_from_subsystem_sysname() if we know the input is the main interface name. */
+
+        r = rtnl_resolve_ifname(NULL, ifname);
+        if (r < 0)
+                return r;
```

Comment:
hmm, what about doing it like this:

1. If the specified name is suitable as primary interface name, then try to suffix it to /sys/class/net/ immediately, and use sd_device_new_from_syspath() on that, and if that works great.
2. If that sd_device_new_from_syspath() call returned -ENODEV or the specified name didn't look like an ifname in the first place but does look like an alternative name, then resolve it via rtnl, and use that.

Net effect: if the specified name is a regular ifname, then we'll generally have success right-away. However we also work with alternative names.